### ISSUE TITLE
Add autofix for object-shorthand 'consistent' option

### ISSUE THREAD

**Issue by chrisvasz** [eslint/eslint#7605]
(edit by @not-an-aardvark: See [this comment](https://github.com/eslint/eslint/issues/7605#issuecomment-260884211) for the updated enhancement proposal.)

**Tell us about your environment**

* **ESLint Version:** 3.10.1
* **Node Version:** 6.7.0
* **npm Version:** 3.10.3

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```json
{
  "parserOptions": {
    "ecmaVersion": 6,
  },
  "rules": {
    "object-shorthand": ["error", "consistent"],
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
let a = 1;
let b = {
  a,
  hello: 'world',
};
```

**What did you expect to happen?**
I expected `eslint --fix` to change the `a` line to `a: a,`.

**What actually happened? Please include the actual, raw output from ESLint.**
`2:9  error  Unexpected mix of shorthand and non-shorthand properties  object-shorthand`

When I change the `object-shorthand` rule from `consistent` to `never`, the autofix works. The autofix also fails for `consistent-as-needed`.

**Comment by platinumazure** [comment]
I think "consistent" and "consistent-as-needed" options try really hard not to make an assumption about which way an object's properties should go (i.e., should they be longhand or shorthand if both are present in an object).

The rule's source [shows](https://github.com/eslint/eslint/blob/master/lib/rules/object-shorthand.js#L192) that for the consistent option, we don't include a fixer at all. So at this point, it seems to be by design.

I wonder if we could consider adding a fixer for "consistent-as-needed" at least, because in that case there's always a correct answer of longhand vs shorthand (if there is one longhand that can't be converted to shorthand, all must be longhand, else all must be shorthand). But for "consistent", I think users might be surprised in some cases by an auto-fixer.

**Comment by not-an-aardvark** [comment]
We do have a fixer for `quote-props`, which has the same options. I think it would be reasonable to never fix to shorthand with `consistent`, and always fix to shorthand for `consistent-as-needed` (if possible).

**Comment by chrisvasz** [comment]
@not-an-aardvark That would certainly solve our use case. Even if we only had a fixer for `consistent-as-needed`, that would save a lot of manual editing.

**Comment by platinumazure** [comment]
@chrisvasz @not-an-aardvark Could one of you please create a rule change proposal using [this template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md) as a guideline? Then we can evaluate the proposal, hopefully accept it, and open the way for a pull request. Thanks!

**Comment by not-an-aardvark** [comment]
**What rule do you want to change?**

`object-shorthand`

**Does this change cause the rule to produce more or fewer warnings?**

The same amount

**How will the change be implemented? (New option, new default behavior, etc.)?**

This will change the default behavior for autofixing.

**Please provide some example code that this change will affect:**

```js
/* eslint object-shorthand: [error, "consistent"] */

({ foo: foo, bar });
```

```js
/* eslint object-shorthand: [error, "consistent-as-needed"] */

({ foo: foo, bar });
```

**What does the rule currently do for this code?**

It reports an error, but does not autofix it.

**What will the rule do after it's changed?**

It will autofix the cases to:

```js
/* eslint object-shorthand: [error, "consistent"] */

({ foo: foo, bar: bar });
```

```js
/* eslint object-shorthand: [error, "consistent-as-needed"] */

({ foo, bar });
```

In other words, the `consistent` option will update the object to use longform for all properties. The `consistent-as-needed` option will use shorthand for all properties if possible, otherwise it will use longform for all properties.

I think this is in line with what the user would expect. The only potential surprise would be with the `consistent` option using longform for all properties. However, if people want to prefer shorthand for all properties in cases like this, they can simply use `consistent-as-needed`. (Note that the [`quote-props`](http://eslint.org/docs/rules/quote-props) fixer already functions like this.)

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
